applications/serial_lte_modem/src/slm_at_socket.c

```diff
@@ -1632,6 +1701,249 @@ static int handle_at_poll(enum at_parser_cmd_type cmd_type, struct at_parser *pa
 	return err;
 }
 
+static void async_poll_thread(void*, void*, void*)
+{
+	int ret;
+	struct zsock_pollfd afds[SLM_FDS_COUNT] = {
+		[0 ... SLM_FDS_COUNT - 1] = {
+			.fd = INVALID_SOCKET
+		}
+	};
+
+	LOG_DBG("Asynchronous polling thread started");
+
+	/* Always poll the EFD socket. */
+	afds[ARRAY_SIZE(afds) - 1].fd = poll_ctx.efd;
+	afds[ARRAY_SIZE(afds) - 1].events = ZSOCK_POLLIN;
+
+	while (poll_ctx.poll_running) {
+		for (int i = 0; i < ARRAY_SIZE(socks); ++i) {
+			/* Skip sockets that are not opened or are disabled for async poll. */
+			if (socks[i].fd == INVALID_SOCKET || socks[i].async_poll.disable) {
+				afds[i].fd = INVALID_SOCKET;
+				continue;
+			}
+			/* Skip sockets that are not in async poll mode. */
+			if (!poll_ctx.poll_all && !socks[i].async_poll.specific) {
+				afds[i].fd = INVALID_SOCKET;
+				continue;
+			}
+			afds[i].fd = socks[i].fd;
+
+			/* atomic_clear returns the value before clear. */
+			afds[i].events |= atomic_clear(&socks[i].async_poll.update_events);
+		}
+
+		ret = zsock_poll(afds, ARRAY_SIZE(afds), -1);
+		if (ret < 0) {
+			LOG_ERR("zsock_poll() failed: %d, exit thread...", -errno);
+			poll_ctx.poll_running = false;
+			break;
+		}
+		for (int i = 0; i < ARRAY_SIZE(afds); ++i) {
+			if (afds[i].fd == INVALID_SOCKET) {
+				continue;
+			}
+
+			if (afds[i].fd == poll_ctx.efd && afds[i].revents) {
+				LOG_DBG("EFD revents: 0x%04x", afds[i].revents);
+				if (afds[i].revents & ZSOCK_POLLIN) {
+					eventfd_t value;
+
+					eventfd_read(poll_ctx.efd, &value);
+					if (value == EFD_POLL) {
+						/* Restart polling. */
+						continue;
+					} else {
+						/* Exit thread. */
+						poll_ctx.poll_running = false;
+						break;
+					}
+				} else {
+					LOG_ERR("EFD revents: 0x%04x",
+						afds[i].revents);
+					break;
+				}
+			}
+
+			if (afds[i].revents) {
+				if (!in_datamode()) {
+					rsp_send_indicate("\r\n#XAPOLL: %d,0x%02x\r\n",
+							  afds[i].fd, afds[i].revents);
+				}
+
+				if (afds[i].revents &
+				    (ZSOCK_POLLERR | ZSOCK_POLLNVAL | ZSOCK_POLLHUP)) {
+					/* Remove socket from poll until closed. */
+					socks[i].async_poll.disable = true;
+				} else {
+					/* Remove POLLIN/POLLOUT from poll, until AT-operation. */
+					afds[i].events &= ~afds[i].revents;
```

Comment:
If socket1 is in datamode, how does this work for other sockets? We don't notify that there is data for them, and we stop polling for incoming data.

This is a complex feature with datamode.